internal/ls/folding.go

```diff
@@ -0,0 +1,536 @@
+	_, sourceFile := l.getProgramAndFile(documentURI)
+	res := l.addNodeOutliningSpans(sourceFile)
+	res = append(res, l.addRegionOutliningSpans(sourceFile)...)
+	sort.Slice(res, func(i, j int) bool {
+		if res[i].StartLine != res[j].StartLine {
+			return res[i].StartLine < res[j].StartLine
+		}
+		if res[i].EndLine != res[j].EndLine {
+			return res[i].EndLine < res[j].EndLine
+		} else if res[i].StartCharacter != nil && res[j].StartCharacter != nil {
+			return *res[i].StartCharacter < *res[j].StartCharacter
+		}
+		return *res[i].EndCharacter < *res[j].EndCharacter
+	})
+	return res
+}
+
+func (l *LanguageService) addNodeOutliningSpans(sourceFile *ast.SourceFile) []*lsproto.FoldingRange {
+	depthRemaining := 40
+	current := 0
+
+	statements := sourceFile.Statements
+	n := len(statements.Nodes)
+	var foldingRange []*lsproto.FoldingRange
+	for current < n {
+		for current < n && !ast.IsAnyImportSyntax(statements.Nodes[current]) {
+			foldingRange = append(foldingRange, visitNode(statements.Nodes[current], depthRemaining, sourceFile, l)...)
+			current++
+		}
+		if current == n {
+			break
+		}
+		firstImport := current
+		for current < n && ast.IsAnyImportSyntax(statements.Nodes[current]) {
+			foldingRange = append(foldingRange, visitNode(statements.Nodes[current], depthRemaining, sourceFile, l)...)
+			current++
+		}
+		lastImport := current - 1
+		if lastImport != firstImport {
+			foldingRangeKind := lsproto.FoldingRangeKindImports
+			foldingRange = append(foldingRange, createFoldingRangeFromBounds(
+				astnav.GetStartOfNode(findChildOfKind(statements.Nodes[firstImport],
+					ast.KindImportKeyword, sourceFile), sourceFile, false /*includeJSDoc*/),
+				statements.Nodes[lastImport].End(),
+				foldingRangeKind,
+				sourceFile,
+				l))
+		}
+	}
+
+	// Visit the EOF Token so that comments which aren't attached to statements are included.
+	foldingRange = append(foldingRange, visitNode(sourceFile.EndOfFileToken, depthRemaining, sourceFile, l)...)
+	return foldingRange
+}
+
+func (l *LanguageService) addRegionOutliningSpans(sourceFile *ast.SourceFile) []*lsproto.FoldingRange {
+	var regions []*lsproto.FoldingRange
+	var out []*lsproto.FoldingRange
+	lineStarts := scanner.GetLineStarts(sourceFile)
+	for _, currentLineStart := range lineStarts {
+		lineEnd := scanner.GetLineEndOfPosition(sourceFile, int(currentLineStart))
+		lineText := sourceFile.Text()[currentLineStart:lineEnd]
+		result := parseRegionDelimiter(lineText)
+		if result == nil || isInComment(sourceFile, int(currentLineStart), nil) != nil {
+			continue
+		}
+		if result.isStart {
+			span := l.createLspPosition(strings.Index(sourceFile.Text()[currentLineStart:lineEnd], "//")+int(currentLineStart), sourceFile)
+			foldingRangeKindRegion := lsproto.FoldingRangeKindRegion
+			collapsedTest := "#region"
+			if result.name != "" {
+				collapsedTest = result.name
+			}
+			regions = append(regions, &lsproto.FoldingRange{
```

Comment:
Okay, there is a subtlety both here and in the original code about how the end position isn't filled in yet.


```suggestion
			// Our spans start out with some initial data.
			// On every `#endregion`, we'll come back to these `FoldingRange`s
			// and fill in their EndLine/EndCharacter.
			regions = append(regions, &lsproto.FoldingRange{
```